src/phylo/iteration.jl

```diff
@@ -0,0 +1,277 @@
+#==================================#
+# Iteration and traversal of trees #
+#==================================#
+
+# Ben J. Ward, 2014.
+
+@Docile.doc """
+`PhylogenyIterator` is an abstract type that defines a family of iterators
+for traversing trees in various ways, including `BreadthFirst`, and `DepthFirst`,
+or from a tip to a root.
+""" ->
+abstract PhylogenyIterator
+
+@Docile.doc """
+A type for implementing breadth-first traversal of any 
+`Phylogeny`.
+
+A `BreathFirst` phylogeny iterator contains:
+
+- A `PhyNode` that the iterator will `start` at.
+""" ->
+immutable BreadthFirst <: PhylogenyIterator
+  start::PhyNode
+  function BreadthFirst(start::PhyNode)
+    if isempty(start)
+      error("Node provided to start iteration is empty!")
+    end
+    return new(start)
+  end
+end
+
+typealias BreadthFirstState Queue{Deque{PhyNode}} 
+
+@Docile.doc """
+Construct a BreadthFirst iterator for a tree.
+
+**Parameters**
+* `x`: A Phylogeny variable you want to traverse in breadth-first order. 
+""" ->
+function BreadthFirst(x::Phylogeny)
+  return BreadthFirst(x.root)
+end
+
+@Docile.doc """
+Start iterating through a `Phylogeny` with a `BreadthFirst` iterator.
+
+**Parameters**
+* `x`: A `BreadthFirst` iterator to start iterating with.
+""" ->
+function Base.start(x::BreadthFirst)
+  state::BreadthFirstState = Queue(PhyNode)
+  enqueue!(state, x.start)
+  return state
+end
+
+@Docile.doc """
+Step to the next node when iterating over a `Phylogeny`.
+
+**Parameters:**
+* `x`: `Breadthfirst` iterator.
+* `state`: A `Queue` that contains the next nodes to be iterated over.
+""" ->
+function Base.next(x::BreadthFirst, state::BreadthFirstState)
+  current::PhyNode = dequeue!(state)
+  for i in current.children
+    enqueue!(state, i)
+  end
+  return current, state
+end
+
+@Docile.doc """
+Check whether the iterator is at the last node in a phylogeny and is therefore
+done iterating.
+
+**Parameters:**
+* `x`: A `BreadthFirst` `PhylogenyIterator`.
+* `state`: a `Queue` containing the nodes that are to be visited imminently.
+""" ->
+function Base.done(x::BreadthFirst, state::BreadthFirstState)
+  return length(state) == 0
+end
+
+@Docile.doc """
+A type for implementing depth-first traversal of any 
+`Phylogeny`.
+
+A `DepthFirst` phylogeny iterator contains:
+
+- A `PhyNode` that the iterator will `start` at.
+""" ->
+immutable DepthFirst <: PhylogenyIterator
+  start::PhyNode
+  function DepthFirst(x::PhyNode)
+    if isempty(x)
```

Comment:
Should it be valid to iterate over an empty tree? Iterating over empty arrays or dicts doesn't result in an error, is this different?